[Bug] [sqlserver] 在 yarn-pre-job 模式找不到 sqlserver 相关驱动

### Search before asking

- [X] I had searched in the [issues](https://github.com/DTStack/chunjun/issues) and found no similar issues.


### What happened

在 yarn-pre-job 模式找不到 sqlserver 相关驱动

```

java.lang.RuntimeException: java.sql.SQLException: No suitable driver found for jdbc:jtds:sqlserver://x.x.x.x:1433;databaseName=test;useLOBs=false
	at com.dtstack.chunjun.util.RetryUtil$Retry.doRetry(RetryUtil.java:125) ~[chunjun-core.jar:?]
	at com.dtstack.chunjun.util.RetryUtil.executeWithRetry(RetryUtil.java:47) ~[chunjun-core.jar:?]
	at com.dtstack.chunjun.connector.jdbc.util.JdbcUtil.getConnection(JdbcUtil.java:112) ~[chunjun-connector-dm-master.jar:?]
	at com.dtstack.chunjun.connector.jdbc.source.JdbcInputFormat.getConnection(JdbcInputFormat.java:809) ~[chunjun-connector-dm-master.jar:?]
	at com.dtstack.chunjun.connector.jdbc.source.JdbcInputFormat.openInternal(JdbcInputFormat.java:107) ~[chunjun-connector-dm-master.jar:?]
	at com.dtstack.chunjun.source.format.BaseRichInputFormat.open(BaseRichInputFormat.java:163) ~[chunjun-core.jar:?]
	at com.dtstack.chunjun.source.DtInputFormatSourceFunction.run(DtInputFormatSourceFunction.java:119) ~[chunjun-core.jar:?]
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:110) ~[flink-dist-1.16.1.jar:1.16.1]
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:67) ~[flink-dist-1.16.1.jar:1.16.1]
	at org.apache.flink.streaming.runtime.tasks.SourceStreamTask$LegacySourceFunctionThread.run(SourceStreamTask.java:333) ~[flink-dist-1.16.1.jar:1.16.1]
Caused by: java.sql.SQLException: No suitable driver found for jdbc:jtds:sqlserver://x.x.x.x:1433;databaseName=test;useLOBs=false
	at java.sql.DriverManager.getConnection(DriverManager.java:689) ~[?:1.8.0_60]
	at java.sql.DriverManager.getConnection(DriverManager.java:208) ~[?:1.8.0_60]
	at com.dtstack.chunjun.connector.jdbc.util.JdbcUtil.lambda$getConnection$1(JdbcUtil.java:113) ~[chunjun-connector-dm-master.jar:?]
	at com.dtstack.chunjun.util.RetryUtil$Retry.call(RetryUtil.java:129) ~[chunjun-core.jar:?]
	at com.dtstack.chunjun.util.RetryUtil$Retry.doRetry(RetryUtil.java:71) ~[chunjun-core.jar:?]
	... 9 more

```



### What you expected to happen

--

### How to reproduce

```

CREATE TABLE source
(
    `a_date` date,
    `b_datetime` timestamp(0),
    `c_nvarchar` string,
    `d_varchar` string,
    `e_datetime2` string,
    `f_decimal` decimal(10,3),
    `g_int` int
) WITH (
      'schema' = 'dbo',
      'scan.fetch-size' = '1024',
      'password' = 'xxx',
      'connector' = 'sqlserver-x',
      'scan.query-timeout' = '600',
      'table-name' = 'test',
      'url' = 'jdbc:jtds:sqlserver://x.x.x.x:1433;databaseName=test;useLOBs=false',
      'username' = 'sa'
      );
CREATE TABLE sink
(
    `a_date` date,
    `b_datetime` timestamp(0),
    `c_nvarchar` string,
    `d_varchar` string,
    `e_datetime2` timestamp(0),
    `f_decimal` decimal(10,3),
    `g_int` int
) WITH (
      'connector' = 'stream-x',
      'print' = 'true'

      );
insert into sink
select
    `a_date` as `a_date`,
    `b_datetime` as `b_datetime`,
    `c_nvarchar` as `c_nvarchar`,
    `d_varchar` as `d_varchar`,
    TO_TIMESTAMP(cast(`e_datetime2` as string))  as `e_datetime2`,
    CAST(`f_decimal` AS decimal(10,3))  as `f_decimal`,
    `g_int` as `g_int`
from source ;


```


### Anything else

_No response_

### Version

master

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/chunjun-connectors/chunjun-connector-sqlserver/src/main/java/com/dtstack/chunjun/connector/sqlserver/table/SqlserverDynamicTableFactory.java
+++ b/chunjun-connectors/chunjun-connector-sqlserver/src/main/java/com/dtstack/chunjun/connector/sqlserver/table/SqlserverDynamicTableFactory.java
@@ -28,7 +28,9 @@ import com.dtstack.chunjun.connector.sqlserver.dialect.SqlserverDialect;
 import com.dtstack.chunjun.connector.sqlserver.sink.SqlserverOutputFormat;
 import com.dtstack.chunjun.connector.sqlserver.source.SqlserverInputFormat;
 
+import org.apache.flink.table.connector.sink.DynamicTableSink;
 import org.apache.flink.table.connector.source.DynamicTableSource;
+import org.apache.flink.table.factories.FactoryUtil;
 
 import org.apache.commons.lang3.StringUtils;
 
@@ -38,6 +40,8 @@ public class SqlserverDynamicTableFactory extends JdbcDynamicTableFactory {
 
     private static final String IDENTIFIER = "sqlserver-x";
 
+    private JdbcConfig jdbcConfig;
+
     @Override
     public String factoryIdentifier() {
         return IDENTIFIER;
@@ -45,6 +49,11 @@ public class SqlserverDynamicTableFactory extends JdbcDynamicTableFactory {
 
     @Override
     protected JdbcDialect getDialect() {
+        if (jdbcConfig != null) {
+            return new SqlserverDialect(
+                    jdbcConfig.isWithNoLock(),
+                    jdbcConfig.getJdbcUrl().startsWith("jdbc:jtds:sqlserver"));
+        }
         return new SqlserverDialect();
     }
 
@@ -60,9 +69,22 @@ public class SqlserverDynamicTableFactory extends JdbcDynamicTableFactory {
 
     @Override
     public DynamicTableSource createDynamicTableSource(Context context) {
+        final FactoryUtil.TableFactoryHelper helper =
+                FactoryUtil.createTableFactoryHelper(this, context);
+        this.jdbcConfig = getSourceConnectionConfig(helper.getOptions());
         Map<String, String> prop = context.getCatalogTable().getOptions();
         prop.put("druid.validation-query", "SELECT 1");
         return super.createDynamicTableSource(context);
+    }
+
+    @Override
+    public DynamicTableSink createDynamicTableSink(Context context) {
+        final FactoryUtil.TableFactoryHelper helper =
+                FactoryUtil.createTableFactoryHelper(this, context);
+        this.jdbcConfig =
+                getSinkConnectionConfig(
+                        helper.getOptions(), context.getCatalogTable().getResolvedSchema());
+        return super.createDynamicTableSink(context);
     }
 
     /** table字段有可能是[schema].[table]格式 需要转换为对应的schema 和 table 字段* */

